File: 1770maxScoreFromMultOperations.py
"""
 @param {number[]} nums
 @param {number[]} multipliers
 @return {number}
"""


class Solution:

    def maximumScore(self, nums, multipliers):
        numslength = len(nums)
        multiplierslength = len(multipliers)

        @lru_cache(2000)
        def dp(l, r):
            numsCurLen = r - l + 1
            iter = numslength - numsCurLen
            if iter > multiplierslength - 1: return 0

            mult = multipliers[iter]
            # have to return value only
            return max(mult * nums[l] + dp(l + 1, r),
                       mult * nums[r] + dp(l, r - 1))

        return dp(0, numslength - 1)

File: 1770maxScoreFromMultOperations.js
/**
 * @param {number[]} nums
 * @param {number[]} multipliers
 * @return {number}
 */
var maximumScore = function (nums, multipliers) {
  const n = nums.length;
  const m = multipliers.length;

  const dp = Array.from({ length: m + 1 }, () => Array(m + 1).fill(0));
  console.log(dp);

  for (let i = m - 1; i > -1; i--) {
    for (let left = i; left > -1; left--) {
      const mult = multipliers[i];
      const right = n - 1 - (i - left);
      dp[i][left] = Math.max(
        mult * nums[left] + dp[i + 1][left + 1],
        mult * nums[right] + dp[i + 1][left]
      );
    }
  }
  return dp[0][0];
};
